oxenc/bt_producer.h

```diff
@@ -201,6 +201,27 @@ class bt_list_producer {
         return ret;
     }
 
+    /// Returns a reference to the `std::string`, when in string-builder mode.  Unlike `str()`, this
+    /// method *can* be used on a subdict/sublist, but always returns a reference to the root
+    /// object's string (unlike `.view()` which just returns the view of the current sub-producer).
+    const std::string& str_ref() {
+        if (auto* p = parent())
+            return p->str_ref();
+        if (auto* s = std::get_if<std::string>(&out))
+            return *s;
+        throw std::logic_error{"Cannot call bt_producer .str_ref() when using an external buffer"};
+    }
+
+    /// Calls `.reserve()` on the underlying std::string, if using string-builder mode.
+    void reserve(size_t new_cap) {
```

Comment:
Should this not use `str_ref` since that already does the parent recursion and string-builder check?

Reply:
str_ref is deliberately `const` so that you can't change our internal string; the `reserve` is the only non-const thing we want to be callable.

Reply:
There could be a private, non-const `str_ref` which both the public `str_ref` as well as `reserve` would use.  Not necessary, but DRY and all that.

Reply:
They are different enough now (reserve doesn't throw anymore).